docs/proposals/admission_control_changes.md

```diff
@@ -0,0 +1,485 @@
+4. r.DeleteCollection ...
+5. admit.Finish
+
+### Connect
+
+#### Current behavior
+
+1. connecter.NewConnectOptions
+2. decode request connect options
+3. admit.Admit
+4. connecter.Connect().ServeHTTP()
+
+#### Proposed behavior
+
+1. connecter.NewConnectOptions
+2. decode request connect options
+1. admit.Validate
+2. connecter.Connect().ServeHTTP()
+
+## Ability to run out-of-tree plugins
+
+Admission plugins are all currently compiled into the apiserver binary. If you
+want to run custom plugins, you have to modify the Kubernetes source code and
+recompile the apiserver. Some examples of custom plugins that an administrator
+might want to run (these are all actual plugins in OpenShift today):
+
+- default restart-never and restart-on-failure pods' `activeDeadlineSeconds`
+  based on either a namespace-level annotation, or the plugin's configured value
+  (although this probably should be part of LimitRange).
+- apply quota at the cluster level instead of per-namespace
+- assign default resource request values relative to user-specified limits for
+  pods to support an administrator-defined overcommit target
+- restrict who is allowed to set the `nodeName` or `nodeSelector` on pods
+- ensure you can't exec or attach to a pod with a privileged security context if
+  you don't have privileges to create a privileged pod
+- supply default values for certain fields on Build Pods when unset (e.g.
+  environment variables)
+- enforce cluster-wide configuration values on all Build Pods (i.e.
+  override user-supplied data if necessary)
+- automatically provision a Jenkins server in a user's namespace if it doesn't
+  exist when a pipeline BuildConfig is created
+
+We would like to have the ability to run a single external plugin at the end of
+each admission phase. This will enable cluster administrators to inject custom
+admission logic into an existing kube-apiserver without having to recompile it.
+
+### OPEN QUESTIONS
+
+#### What happens if the external admission plugin is not available?
+
+Let's imagine that I have a Kubernetes cluster up and running. Now I want to run
+OpenShift on top of it, so I deploy its various pods to the cluster, presumably
+as Deployment resources. Next, I need to add OpenShift as an external admission
+plugin for the kube-apiserver pods and restart them.
+
+Fast forward a bit - now I have everything working. Then let's imagine that my
+openshift-apiserver pods fail for whatever reason. What happens to
+kube-apiserver requests with respect to admission? With openshift-apiserver
+down, any attempt by the kube-apiserver to invoke the external admission plugin
+will fail.
+
+Do we ignore an external admission plugin being unreachable and allow requests
+to proceed ("fail open")? I don't think this is viable. Presumably the external
+admission plugin is there for a reason, and if it is unable to execute, it opens
+the possibility for inconsistent or incorrect data to enter the cluster.
```

Comment:
> Suggestion: the admission plugin needs to run in a namespace which itself doesn't require the admission plugin to be run. So admins can get the admission plugin up and running without it needing to already be running. This requires an on/off setting per-namespace, though.

Doesn't require or doesn't allow?  It seems like it may get confusing to end up with different results depending on the state of your cluster.  Clayton and I discussed a similar config option for image policy rules: have a privileged namespace that never runs that admission plugin so you can run the openshift registry to boostrap yourself.